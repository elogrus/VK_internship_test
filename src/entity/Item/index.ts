import { IItem, IItemList, IOwner } from "./types/Item";
import { ItemList } from "./ui/ItemList/ItemList";
export { IItem, IItemList, IOwner, ItemList };

