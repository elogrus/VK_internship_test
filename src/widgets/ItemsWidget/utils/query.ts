import { GithubSorts } from "shared/consts/GithubSorts";

interface IQuery {
    ITEMS_ON_PAGE: number;
    getURL: (query: string, page: number, sort: GithubSorts) => string;
}

export const Query: IQuery = {
    ITEMS_ON_PAGE: 10,
    getURL: (query, page, sort) =>
        `https://api.github.com/search/repositories?q=${query}&page=${page}&per_page=${Query.ITEMS_ON_PAGE}&sort=${sort}`,
};
