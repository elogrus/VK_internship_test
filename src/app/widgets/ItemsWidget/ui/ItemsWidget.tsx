import { ItemList } from "entity/Item";
import * as cls from "./ItemsWidget.module.scss";

export const ItemsWidget = () => {
    return (
        <div className={cls.ItemsWidget}>
            {/* <span>Результат: {items.total_count}</span> */}
            <ItemList />
        </div>
    );
};
