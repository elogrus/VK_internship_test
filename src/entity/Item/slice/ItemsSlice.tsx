import type { PayloadAction } from "@reduxjs/toolkit";
import { createSlice } from "@reduxjs/toolkit";
import { IItem, IItemList } from "entity/Item/types/Item";

const initialState = {
    items: [],
    total_count: -1,
} satisfies IItemList as IItemList;

const ItemsSlice = createSlice({
    name: "Items",
    initialState,
    reducers: {
        setTotalCount(state, action: PayloadAction<number>) {
            state.total_count = action.payload;
            return state;
        },
        addItems(state, action: PayloadAction<IItem[]>) {
            const items = action.payload;
            state.items.push(...items);
            return state;
        },
        setItems(state, action: PayloadAction<IItem[]>) {
            const items = action.payload;
            state.items = items;
            return state;
        },
        editItem(state, action: PayloadAction<[id: number, newItem: IItem]>) {
            const [id, newItem] = action.payload;
            const itemIndex = state.items.findIndex((item) => item.id === id);
            state.items[itemIndex] = newItem;
            return state;
        },
        removeItem(state, action: PayloadAction<number>) {
            const id = action.payload;
            const itemIndex = state.items.findIndex((item) => item.id === id);
            state.items.splice(itemIndex, 1);
            return state;
        },
    },
});

export const { addItems, setItems, editItem, removeItem, setTotalCount } =
    ItemsSlice.actions;
export default ItemsSlice.reducer;
