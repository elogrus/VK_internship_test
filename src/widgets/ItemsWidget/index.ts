import { ItemsWidgetAsync } from "./ui/ItemsWidget.async";
export { ItemsWidgetAsync as ItemsWidget };
