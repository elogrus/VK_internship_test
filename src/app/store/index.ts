import { configureStore } from "@reduxjs/toolkit";
import { useDispatch, useSelector, useStore } from "react-redux";
import ItemsSlice from "../../entity/Item/slice/ItemsSlice";

export const store = configureStore({
    reducer: {
        GithubItems: ItemsSlice,
    },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type AppStore = typeof store;

export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export const useAppSelector = useSelector.withTypes<RootState>();
export const useAppStore = useStore.withTypes<AppStore>();
