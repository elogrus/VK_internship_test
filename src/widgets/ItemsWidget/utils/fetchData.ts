import { useAppDispatch } from "app/store";
import { IItemList } from "entity/Item";
import { addItems, setTotalCount } from "entity/Item/slice/ItemsSlice";
import { MutableRefObject } from "react";
import { IQueryRef } from "../ui/ItemsWidget";
import { Query } from "./query";

export const fetchData = async (
    queryRef: MutableRefObject<IQueryRef>,
    setIsLoading: (val: boolean) => void,
    dispatch: ReturnType<typeof useAppDispatch>
) => {
    setIsLoading(true);
    try {
        queryRef.current.blocked = true;
        const response = await fetch(
            Query.getURL(
                queryRef.current.queryString,
                queryRef.current.page,
                queryRef.current.sort
            )
        );
        if (response.status >= 400) throw Error(response.statusText);
        const data: IItemList = await response.json();
        queryRef.current.page += 1;
        dispatch(addItems(data.items));
        dispatch(setTotalCount(data.total_count));
    } catch (error) {
        queryRef.current.blocked = true;
    } finally {
        queryRef.current.blocked = false;
        setIsLoading(false);
    }
};
