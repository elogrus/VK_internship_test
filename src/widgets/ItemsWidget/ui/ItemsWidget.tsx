import { useAppDispatch, useAppSelector } from "app/store";
import { IItem } from "entity/Item";
import { editItem, removeItem, setItems } from "entity/Item/slice/ItemsSlice";
import {
    MouseEventHandler,
    useCallback,
    useEffect,
    useRef,
    useState,
} from "react";
import { GithubSorts } from "shared/consts/GithubSorts";
import { fetchData } from "../utils/fetchData";
import * as cls from "./ItemsWidget.module.scss";
import { ContentPart } from "./parts/ContentPart/ContentPart";
import { FormPart } from "./parts/FormPart/FormPart";
import { Alert } from "@mui/material";

export interface IQueryRef {
    queryString: string;
    sort: GithubSorts;
    page: number;
    blocked: boolean;
}

const ItemsWidget = () => {
    const dispatch = useAppDispatch();
    const [isLoading, setIsLoading] = useState(false);
    const ObservantRef = useRef();
    const list = useAppSelector((state) => state.GithubItems);

    // Состояние после нажатия кнопки "найти", не меняется до следующего нажатия
    const queryRef = useRef<IQueryRef>({
        queryString: "javascript",
        sort: GithubSorts.STARS,
        page: 1,
        blocked: false,
    });

    // Всегда актуальное состояние инпутов
    const formRef = useRef<IQueryRef>({
        queryString: "javascript",
        sort: GithubSorts.STARS,
        page: 1,
        blocked: false,
    });

    useEffect(() => {
        const observer = new IntersectionObserver(async (entries) => {
            if (!queryRef.current.blocked && entries[0].isIntersecting) {
                fetchData(queryRef, setIsLoading, dispatch);
            }
        });
        observer.observe(ObservantRef.current);
    }, []);

    const onSend: MouseEventHandler<HTMLButtonElement> = (e) => {
        dispatch(setItems([]));
        queryRef.current = { ...queryRef.current, ...formRef.current };
        fetchData(queryRef, setIsLoading, dispatch);
    };

    const onEditItem = useCallback((id: number, newItem: IItem) => {
        dispatch(editItem([id, newItem]));
    }, []);

    const onDeleteItem = useCallback((id: number) => {
        dispatch(removeItem(id));
    }, []);

    return (
        <div className={cls.ItemsWidget}>
            <FormPart
                onInputChange={(e) =>
                    (formRef.current.queryString = e.currentTarget.value)
                }
                onSelectChange={(e) =>
                    (formRef.current.sort = e.target.value as GithubSorts)
                }
                onSend={onSend}
                queryRef={queryRef}
            />
            <ContentPart
                list={list}
                onEditItem={onEditItem}
                onDeleteItem={onDeleteItem}
                ObservantRef={ObservantRef}
                isLoading={isLoading}
            />
        </div>
    );
};

export default ItemsWidget;
