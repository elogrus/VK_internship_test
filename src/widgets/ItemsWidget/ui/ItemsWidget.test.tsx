import {
    act,
    fireEvent,
    render,
    screen,
    waitFor,
} from "@testing-library/react";
import * as reduxHooks from "app/store";
import { IItemList } from "entity/Item";
import * as actions from "entity/Item/slice/ItemsSlice";
import { getTestItem } from "entity/Item/testHelpers/getItem";
import { intersect } from "setupTests";
import { GithubSorts } from "shared/consts/GithubSorts";
import { Query } from "../utils/query";
import ItemsWidget from "./ItemsWidget";

const ItemList: IItemList = {
    total_count: 1000,
    items: [
        getTestItem(1, "name1"),
        getTestItem(2, "name2"),
        getTestItem(3, "name3"),
    ],
};

describe("ItemsWidget", () => {
    jest.spyOn(reduxHooks, "useAppSelector").mockReturnValue(ItemList);
    const dispatch = jest.fn();
    jest.spyOn(reduxHooks, "useAppDispatch").mockReturnValue(dispatch);

    const renderWidget = () => render(<ItemsWidget />);

    it("renders", () => {
        renderWidget();
        const root = screen.getByTestId("ItemsWidget_root");
        expect(root).toBeInTheDocument();
        expect(root).toMatchSnapshot();
    });

    it("show items", () => {
        renderWidget();
        expect(screen.queryAllByTestId("Item_root").length).toBe(3);
    });

    it("show progress", async () => {
        global.fetch = jest.fn(() => new Promise(jest.fn()));
        renderWidget();
        await act(
            async () =>
                await intersect(
                    screen.getByTestId("ItemsWidget_observant"),
                    true
                )
        );
        expect(
            screen.queryByTestId("ItemsWidget_progress")
        ).toBeInTheDocument();
    });

    it("show progress on send click", async () => {
        // @ts-ignore
        global.fetch = jest.fn(() =>
            Promise.resolve({
                json: () => Promise.resolve(ItemList),
                status: 200,
                statusText: "OK",
            })
        );
        renderWidget();
        await act(
            async () =>
                await intersect(
                    screen.getByTestId("ItemsWidget_observant"),
                    true
                )
        );
        global.fetch = jest.fn(() => new Promise(jest.fn()));
        fireEvent.click(screen.getByTestId("ItemsWidget_sendButton"));
        expect(
            screen.queryByTestId("ItemsWidget_progress")
        ).toBeInTheDocument();
    });

    it("show no-items-text", () => {
        jest.spyOn(reduxHooks, "useAppSelector").mockReturnValue({
            total_count: 0,
            items: [],
        });
        renderWidget();
        expect(screen.queryByText("Записи не найдены")).toBeInTheDocument();
    });

    it("show items-ran-out-text #1 (less length)", () => {
        jest.spyOn(reduxHooks, "useAppSelector").mockReturnValue({
            ...ItemList,
            total_count: 1,
        });
        renderWidget();
        expect(screen.queryByText("Записи закончились")).toBeInTheDocument();
    });

    it("show items-ran-out-text #2 (equal length)", () => {
        jest.spyOn(reduxHooks, "useAppSelector").mockReturnValue({
            ...ItemList,
            total_count: 3,
        });
        renderWidget();
        expect(screen.queryByText("Записи закончились")).toBeInTheDocument();
    });

    it("fetch data", async () => {
        // @ts-ignore
        global.fetch = jest.fn(() =>
            Promise.resolve({
                json: () => Promise.resolve(ItemList),
                status: 200,
                statusText: "OK",
            })
        );
        renderWidget();
        const select = screen
            .getByTestId("ItemsWidget_select")
            .querySelector("input");
        const input = screen
            .getByTestId("ItemsWidget_input")
            .querySelector("input");
        const sendButton = screen.getByTestId("ItemsWidget_sendButton");

        fireEvent.change(select, { target: { value: GithubSorts.FORKS } });
        fireEvent.change(input, { target: { value: "paramparam" } });
        fireEvent.click(sendButton);

        await waitFor(() => {
            expect(fetch).toHaveBeenCalledTimes(1);
            expect(fetch).toHaveBeenCalledWith(
                Query.getURL("paramparam", 1, GithubSorts.FORKS)
            );

            expect(dispatch).toHaveBeenCalledWith(
                actions.addItems(ItemList.items)
            );
            expect(dispatch).toHaveBeenCalledWith(
                actions.setTotalCount(ItemList.total_count)
            );
        });
    });

    it("on change item", () => {
        jest.spyOn(reduxHooks, "useAppSelector").mockReturnValue({
            total_count: 1000,
            items: [getTestItem(1, "name1")],
        });
        renderWidget();
        fireEvent.click(screen.queryByTestId("Item_editButton"));
        fireEvent.change(screen.getByPlaceholderText("Название"), {
            target: { value: "anotherName" },
        });
        fireEvent.click(screen.queryByTestId("Item_saveButton"));

        expect(dispatch).toHaveBeenCalledWith(
            actions.editItem([1, getTestItem(1, "anotherName")])
        );
    });

    it("on delete item", () => {
        jest.spyOn(reduxHooks, "useAppSelector").mockReturnValue({
            total_count: 1000,
            items: [getTestItem(1, "name1"), getTestItem(2, "name1")],
        });
        renderWidget();
        fireEvent.click(screen.queryAllByTestId("Item_deleteButton")[1]);

        expect(dispatch).toHaveBeenCalledWith(actions.removeItem(2));
    });
});
