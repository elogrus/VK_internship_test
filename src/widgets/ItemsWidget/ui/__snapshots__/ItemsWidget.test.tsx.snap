// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemsWidget renders 1`] = `
<div
  data-testid="ItemsWidget_root"
>
  <form
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1gtchvp-MuiPaper-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <div
      class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
      data-testid="ItemsWidget_select"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-113d811-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r0:"
        id=":r0:-label"
      >
        Сортировать по
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <div
          aria-controls=":r1:"
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby=":r0:-label :r0:"
          class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-w76bbz-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          role="combobox"
          tabindex="0"
        >
          Звездам
        </div>
        <input
          aria-hidden="true"
          aria-invalid="false"
          class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
          tabindex="-1"
          value="stars"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-lohd6h-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-w1u3ce"
          >
            <span>
              Сортировать по
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth css-1i4yu0n-MuiInputBase-root"
      data-testid="ItemsWidget_input"
    >
      <input
        class="MuiInputBase-input css-yimnyd-MuiInputBase-input"
        type="text"
        value="javascript"
      />
    </div>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1u171q0-MuiButtonBase-root-MuiIconButton-root"
      data-testid="ItemsWidget_sendButton"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="SendIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M2.01 21 23 12 2.01 3 2 10l15 2-15 2z"
        />
      </svg>
    </button>
  </form>
  <div
    class="MuiBox-root css-0"
  >
    <h6
      class="MuiTypography-root MuiTypography-h6 css-1ie3cgg-MuiTypography-root"
    >
      Всего найдено: 
      1000
       записей
    </h6>
    <ul
      class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
      data-testid="ItemList_root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
        data-testid="Item_root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1orsvah-MuiTypography-root-MuiLink-root"
              data-testid="Item_owner"
              href="ownerHtmlUrl"
              rel="noreferrer"
              target="_blank"
            >
              <div
                aria-label="login"
                class="MuiAvatar-root MuiAvatar-circular css-fmeqll-MuiAvatar-root"
                data-testid="Item_avatar"
              >
                <img
                  class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
                  src="avatarUrl"
                />
              </div>
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-subtitle1 css-1i3tsf3-MuiTypography-root"
              >
                login
              </h6>
            </a>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineNone css-18rqtra-MuiTypography-root-MuiLink-root"
              data-testid="Item_project"
              href="htmlUrl"
              rel="noreferrer"
              target="_blank"
            >
              <h5
                class="MuiTypography-root MuiTypography-h5 css-16xl4zq-MuiTypography-root"
              >
                name1
              </h5>
              <span
                class="MuiTypography-root MuiTypography-caption css-mpaqru-MuiTypography-root"
              >
                some desc
              </span>
            </a>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-1ypmsg6-MuiTypography-root"
            >
              <div
                data-testid="Item_colorCircle"
                style="background-color: rgb(227, 76, 38);"
              />
              HTML
            </h6>
          </div>
        </div>
        <div
          class="MuiCardActions-root MuiCardActions-spacing css-1q4nm6f-MuiCardActions-root"
          data-testid="Item_buttons"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInfo MuiIconButton-sizeMedium css-1mjraff-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="Item_editButton"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
              />
            </svg>
          </button>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorError MuiIconButton-sizeMedium css-lky15p-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="Item_deleteButton"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
        data-testid="Item_root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1orsvah-MuiTypography-root-MuiLink-root"
              data-testid="Item_owner"
              href="ownerHtmlUrl"
              rel="noreferrer"
              target="_blank"
            >
              <div
                aria-label="login"
                class="MuiAvatar-root MuiAvatar-circular css-fmeqll-MuiAvatar-root"
                data-testid="Item_avatar"
              >
                <img
                  class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
                  src="avatarUrl"
                />
              </div>
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-subtitle1 css-1i3tsf3-MuiTypography-root"
              >
                login
              </h6>
            </a>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineNone css-18rqtra-MuiTypography-root-MuiLink-root"
              data-testid="Item_project"
              href="htmlUrl"
              rel="noreferrer"
              target="_blank"
            >
              <h5
                class="MuiTypography-root MuiTypography-h5 css-16xl4zq-MuiTypography-root"
              >
                name2
              </h5>
              <span
                class="MuiTypography-root MuiTypography-caption css-mpaqru-MuiTypography-root"
              >
                some desc
              </span>
            </a>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-1ypmsg6-MuiTypography-root"
            >
              <div
                data-testid="Item_colorCircle"
                style="background-color: rgb(227, 76, 38);"
              />
              HTML
            </h6>
          </div>
        </div>
        <div
          class="MuiCardActions-root MuiCardActions-spacing css-1q4nm6f-MuiCardActions-root"
          data-testid="Item_buttons"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInfo MuiIconButton-sizeMedium css-1mjraff-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="Item_editButton"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
              />
            </svg>
          </button>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorError MuiIconButton-sizeMedium css-lky15p-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="Item_deleteButton"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
        data-testid="Item_root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1orsvah-MuiTypography-root-MuiLink-root"
              data-testid="Item_owner"
              href="ownerHtmlUrl"
              rel="noreferrer"
              target="_blank"
            >
              <div
                aria-label="login"
                class="MuiAvatar-root MuiAvatar-circular css-fmeqll-MuiAvatar-root"
                data-testid="Item_avatar"
              >
                <img
                  class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
                  src="avatarUrl"
                />
              </div>
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-subtitle1 css-1i3tsf3-MuiTypography-root"
              >
                login
              </h6>
            </a>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineNone css-18rqtra-MuiTypography-root-MuiLink-root"
              data-testid="Item_project"
              href="htmlUrl"
              rel="noreferrer"
              target="_blank"
            >
              <h5
                class="MuiTypography-root MuiTypography-h5 css-16xl4zq-MuiTypography-root"
              >
                name3
              </h5>
              <span
                class="MuiTypography-root MuiTypography-caption css-mpaqru-MuiTypography-root"
              >
                some desc
              </span>
            </a>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-1ypmsg6-MuiTypography-root"
            >
              <div
                data-testid="Item_colorCircle"
                style="background-color: rgb(227, 76, 38);"
              />
              HTML
            </h6>
          </div>
        </div>
        <div
          class="MuiCardActions-root MuiCardActions-spacing css-1q4nm6f-MuiCardActions-root"
          data-testid="Item_buttons"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInfo MuiIconButton-sizeMedium css-1mjraff-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="Item_editButton"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
              />
            </svg>
          </button>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorError MuiIconButton-sizeMedium css-lky15p-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="Item_deleteButton"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
              />
            </svg>
          </button>
        </div>
      </div>
    </ul>
    <div
      data-testid="ItemsWidget_observant"
    />
  </div>
</div>
`;
